lib/fluent/plugin/sd_dns.rb

```diff
@@ -0,0 +1,135 @@
+      include PluginHelper::Mixin
+
+      Plugin.register_sd('dns', self)
+
+      helpers :timer
+
+      desc 'Ignore DNS related error or not'
+      config_param :ignore_dns_error, :bool, default: false
+      desc 'Interval to refresh service list'
+      config_param :interval, :time, default: nil
+
+      config_section :entry, param_name: :entry_configs, required: true, multi: true do
+        desc 'Hostname for service lookup'
+        config_param :host, :string
+        desc 'Port of service'
+        config_param :port, :integer
+        desc 'Specify the address family for returned address'
+        config_param :address_family, default: nil do |val|
+          case val.downcase
+          when 'ipv4'
+            Socket::Constants::AF_INET
+          when 'ipv6'
+            Socket::Constants::AF_INET6
+          else
+            raise Fluent::ConfigError, "Unsupported parameter value: #{val}"
+          end
+        end
+        desc 'The shared_key for authentication per host'
+        config_param :shared_key, :string, default: nil, secret: true
+        desc 'The username for authentication per host'
+        config_param :username, :string, default: ''
+        desc 'The password for authentication per host'
+        config_param :password, :string, default: '', secret: true
+        desc 'The load balancing weight per host'
+        config_param :weight, :integer, default: 60
+      end
+
+      def configure(conf)
+        super
+        @services = fetch_services_from_dns
+      end
+
+      def start(queue)
+        if @interval
+          timer_execute(:"sd_dns_refresh", @interval) do
+            refresh_service_list(queue)
+          end
+        else
+          refresh_service_list(queue)
+        end
+
+        super()
+      end
+
+      private
+
+      def refresh_service_list(queue)
+        old_services = @services
+        new_services = begin
+                         fetch_services_from_dns
+                       rescue => e
+                         @log.error("sd_dns: #{e}")
+                         return
+                       end
+
+        if new_services.nil? || new_services.empty?
+          return
+        end
+
+        diff = []
+        join = new_services - old_services
+        # Need service_in first to guarantee that server exist at least one all time.
+        join.each do |j|
+          diff << ServiceDiscovery.service_in_msg(j)
+        end
+
+        drain = old_services - new_services
+        drain.each do |d|
+          diff << ServiceDiscovery.service_out_msg(d)
+        end
+
+        @services = new_services
+
+        diff.each do |a|
+          queue.push(a)
+        end
+      end
+
+      def fetch_services_from_dns
+        service_list = []
+        @entry_configs.each { |conf|
+          ai_list = begin
+                      Socket.getaddrinfo(conf.host, conf.port, conf.address_family, Socket::SOCK_STREAM)
+                    rescue => e
```

Comment:
Should we rescue all exceptions or specific ones?

Reply:
I think RuntimeException is enough. If other exceptions happens, it seems the bug of code.

Reply:
Oh, I got it. LGTM.
https://docs.ruby-lang.org/ja/2.3.0/class/StandardError.html